### PR TITLE
Support COPY ON SEGMENT command

### PR DESCRIPTION
**Pull request by adam8157**
Support COPY statement that exports the table directly from segment
to local file parallelly.

This commit adds a keyword "on segment" to save the copied file on
"segment" instead of on "master".

Two place holders are used, which are "<SEG_DATA_DIR>" and "<SEGID>"
and will be replced to segment datadir and segment id.

E.g.

```
COPY tbl TO '/tmp/<SEG_DATA_DIR>filename<SEGID>.txt' ON SEGMENT;
```

Signed-off-by: Yuan Zhao <yuzhao@pivotal.io>
Signed-off-by: Haozhou Wang <hawang@pivotal.io>
Signed-off-by: Adam Lee <ali@pivotal.io>

### CODE REVIEW

File: src/backend/commands/copy.c [hunk 1]
```diff
@@ -465,6 +465,16 @@ CopySendEndOfRow(CopyState cstate)
 				ereport(ERROR,
 						(errcode_for_file_access(),
 						 errmsg("could not write to COPY file: %m")));
+
+			/* Send "\n" to QD for "processed" line number counting */
+			if (cstate->on_segment && Gp_role == GP_ROLE_EXECUTE)
+			{
+				if (!cstate->not_count_line) /* not the csv header */
```
Comment (hsyuan) [inline src/backend/commands/copy.c]: does `not_count_line` mean ignore csv header? The var name seems vague.
BTW, double negative is hard on brain. Can we avoid it?

File: src/backend/commands/copy.c [hunk 2]
```diff
@@ -1267,7 +1285,37 @@ DoCopyInternal(const CopyStmt *stmt, const char *queryString, CopyState cstate)
 						 "psql's \\copy command also works for anyone.")));
 
 	cstate->copy_dest = COPY_FILE;		/* default */
-	cstate->filename = (Gp_role == GP_ROLE_EXECUTE ? NULL : stmt->filename); /* QE COPY always uses STDIN */
+    if (Gp_role == GP_ROLE_EXECUTE)
```
Comment (hsyuan) [inline src/backend/commands/copy.c]: indent

File: src/backend/commands/copy.c [hunk 3]
```diff
@@ -2228,7 +2317,7 @@ CopyTo(CopyState cstate)
 	if (cstate->binary)
 	{
 		/* binary header should not be sent in execute mode. */
-		if (Gp_role != GP_ROLE_EXECUTE)
+		if (Gp_role != GP_ROLE_EXECUTE || (Gp_role == GP_ROLE_EXECUTE && cstate->on_segment ))
```
Comment (hsyuan) [inline src/backend/commands/copy.c]: Do we need `Gp_role == GP_ROLE_EXECUTE`? It seems redundant here.

File: src/backend/commands/copy.c [hunk 4]
```diff
@@ -2251,7 +2340,7 @@ CopyTo(CopyState cstate)
 		if (cstate->header_line)
 		{
 			/* header should not be printed in execute mode. */
-			if (Gp_role != GP_ROLE_EXECUTE)
+			if (Gp_role != GP_ROLE_EXECUTE || (Gp_role == GP_ROLE_EXECUTE && cstate->on_segment ))
```
Comment (hsyuan) [inline src/backend/commands/copy.c]: same with above

File: src/backend/commands/copy.c [hunk 5]
```diff
@@ -1286,25 +1334,32 @@ DoCopyInternal(const CopyStmt *stmt, const char *queryString, CopyState cstate)
 		{
 			mode_t		oumask; /* Pre-existing umask value */
 			struct stat st;
+			char* filename = cstate->filename;
```
Comment (hsyuan) [inline src/backend/commands/copy.c]: in Postgres style, it should be `char   *filename`

File: src/backend/commands/copy.c [hunk 6]
```diff
@@ -1897,7 +1970,14 @@ static void CopyToCreateDispatchCommand(CopyState cstate,
 			appendStringInfo(cdbcopy_cmd, ")");
 	}
 
-	appendStringInfo(cdbcopy_cmd, " TO STDOUT WITH");
+	if (cstate->on_segment)
+	{
+		appendStringInfo(cdbcopy_cmd, " TO '%s' WITH ON SEGMENT", cstate->filename);
```
Comment (jmcatamney) [inline src/backend/commands/copy.c]: Should the end of the string here be "ON SEGMENT WITH" instead of "WITH ON SEGMENT", to mirror "TO STDOUT WITH"?

**Comment by jmcatamney** [comment]
Will <SEG_DATA_DIR> be replaced by the entire filepath of the segment data directory, or just the data directory itself?  For example, if a cluster's segment 0 is at /data/primary/gpseg0 and a user executes the sample COPY statement in the description above, will the file on segment 0 be at /tmp/data/primary/gpseg0/filename0.txt or will it be at /tmp/gpseg0/filename0.txt?

From the example given, it looks like the second one is true.  If that's the case, how would a user execute a command to copy a table out into the segment data directories themselves if they don't all share a common root directory (e.g. if the segments are at /data1/primary/gpseg0, /data1/primary/gpseg1, /data2/primary/gpseg2, and /data2/primary/gpseg3)?

**Comment by lij55** [comment]
It should be "/tmp/data/primary/gpseg0/filename0.txt".

**Comment by khuddlefish** [comment]
From our initial testing, it seems like <SEG_DATA_DIR> is just filling in the name of the directory (gpseg0 in Jamie's example). For our use case, we do need it to be the full path. For example, if the segment data directory is `/data/primary/gpseg0` then 
`COPY tbl TO '<SEG_DATA_DIR>/filename<SEGID>.txt' ON SEGMENT;` should put the file in `/data/primary/gpseg0/filename0.txt` and not fail because it doesn't recognize the <SEG_DATA_DIR> as an absolute path. Otherwise, this is looking good. Thank you for your help!

**Comment by HaozhouWang** [comment]
@lij55 @khuddlefish @hsyuan @jmcatamney Thank you very much for review! 
We have fixed based on your comments.

**Comment by dyozie** [comment]
Is this only used internally, or are changes needed to the end-user [COPY}(https://github.com/greenplum-db/gpdb/blob/master/gpdb-doc/dita/ref_guide/sql_commands/COPY.xml) documentation?

**Comment by HaozhouWang** [comment]
@dyozie end-user documentation is needed

**Comment by cjcjameson** [comment]
Oh funny -- y'all updated the docs in `docs/` but @dyozie is suggesting that changes are needed for `gpdb-doc/`? That's super confusing all around. What should we make of it? Thank you for trying to submit the docs change as part of the PR -- that's great and we should keep doing it.

**Comment by adam8157** [comment]
@cjcjameson the doc story is already created in @dyozie 's tracker, which is for the end-user documentation. The manual I updated is not what he asked.

**Comment by cjcjameson** [comment]
Agreed -- raising it on the similar thread on the dev list

**Comment by janebeckman** [comment]
I'm working on preliminary updates to the COPY page for end-user docs per the tracker mentioned above.

**Pull request merged**